### PR TITLE
Refactor spirv-builder to build codegen and sysroot itself

### PR DESCRIPTION
**Pull request by XAMPPRocky**
This PR replaces building `rustc_codegen_spirv` and `spirv-unknown-unknown` with cargo in spirv-builder, with building each as a separate step. Currently this is actually slower as the tests rebuild those artefacts for each invocation, but I'm planning on finishing up the ability to use pre-built artefacts for tests before marking this PR as not a draft. 

Huge thanks to @diananites who added a lot of functionality to `cargo-sysroot` to make this possible. 🎉

### CODE REVIEW

File: examples/runners/ash/build.rs [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
+use std::path::PathBuf;
+
+use spirv_builder::{Source, SourceKind, SpirvBuilder};
 
 fn main() -> Result<(), Box<dyn Error>> {
+    let out_dir = PathBuf::from(std::env::var("OUT_DIR")?);
     // This will set the env var `sky-shader.spv` to a spir-v file that can be include!()'d
-    SpirvBuilder::new("../../shaders/sky-shader").build()?;
+    SpirvBuilder::new(
+        "../../shaders/sky-shader",
```
Comment (DJMcNab) [inline examples/runners/ash/build.rs:10]: I get slightly more luck on windows doing 
```rust
let out_dir = std::fs::canonicalize(std::env::var("OUT_DIR")?)?;
```
here  
Although now I fail at calling link.exe

**Comment by DJMcNab** [comment]
```diff
diff --git a/crates/spirv-builder/src/lib.rs b/crates/spirv-builder/src/lib.rs
index 5ae09739b..501a76947 100644
--- a/crates/spirv-builder/src/lib.rs
+++ b/crates/spirv-builder/src/lib.rs
@@ -312,6 +312,8 @@ impl fmt::Display for SpirvToolsFeatures {
                 Self::Auto => {
                     if std::process::Command::new("spirv-val")
                         .args(&["-h"])
+                        .stderr(Stdio::null())
+                        .stdout(Stdio::null())
                         .status()
                         .map_or(false, |s| s.success())
                     {

```

**Comment by DJMcNab** [comment]
https://github.com/rust-lang/cargo/pull/8874#issuecomment-733397606

Is relevant and might help solve our windows woes.

**Comment by XAMPPRocky** [comment]
Going to close this for now, as the artifact dependency RFC for cargo was merged, and if we did want to go down this route I'd need to rebase it.

**Pull request closed without merging**